### PR TITLE
Modify runtime descriptors to support ADR 0004

### LINKED ISSUE #3450: Implement ADR 4 (part 1, runtime descriptors)
Runtime descriptors, runtime parameters, and the genesis document structures should be changed as described in https://github.com/oasisprotocol/oasis-core/blob/master/docs/adr/0004-runtime-governance.md.
Additional changes to support the new structures are required in the registry, staking, and scheduler.

Old descriptors are accepted and converted into the new format at genesis time, but all new runtime registrations must use the new descriptor format.

After this issue, as well as issues #3429 and #3430 are complete, the rest of the ADR (messages) can be implemented.  A separate issue will be made for that later.

### PR DESCRIPTION
**Pull request by abukosek**
Closes #3450.

TODO:
- [x] fix e2e tests.
- [x] fix unit tests.
- [x] remove `SignedRuntime` entirely.
- [x] double-check that `oasis-node debug fix-genesis` command does the right thing.
- [x] write a few changelog fragments.
- [x] when 3511 is merged, fix the TODO in `go/consensus/tendermint/apps/registry/transactions.go`.
- [x] update docs.
- [x] add `enable_runtime_governance_models` consensus parameter & tests for it.
- [x] add more tests.

In a separate PR:
- add `update_runtime` runtime message support & e2e tests.

### CODE REVIEW

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 1]
```diff
@@ -241,12 +242,71 @@ func (app *registryApplication) registerNode( // nolint: gocyclo
 
 	// Check runtime's whitelist.
 	for _, rt := range paidRuntimes {
-		if rt.AdmissionPolicy.EntityWhitelist != nil && !rt.AdmissionPolicy.EntityWhitelist.Entities[newNode.EntityID] {
-			ctx.Logger().Error("RegisterNode: node's entity not in a runtime's whitelist",
-				"entity", newNode.EntityID,
-				"runtime", rt.ID,
-			)
-			return registry.ErrForbidden
+		if rt.AdmissionPolicy.EntityWhitelist != nil {
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:245]: Maybe invert some checks (with continue statements) and get rid of some indentation.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 2]
```diff
@@ -241,12 +242,71 @@ func (app *registryApplication) registerNode( // nolint: gocyclo
...
+						n, grr := state.Node(ctx, nID)
+						if grr != nil {
+							ctx.Logger().Error("RegisterNode: failed to query node",
+								"err", grr,
+								"node", nID,
+							)
+							return grr
+						}
+
+						if n.HasRoles(role) {
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:297]: This should only count nodes that are registering for the same runtime.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 3]
```diff
@@ -241,12 +242,71 @@ func (app *registryApplication) registerNode( // nolint: gocyclo
...
+						// No such role found in whitelist.
+						return registry.ErrForbidden
+					}
+					if maxNodes == 0 {
+						// No nodes of this type are allowed.
+						return registry.ErrForbidden
+					}
+
+					// Count existing nodes owned by entity.
+					ent, grr := state.Entity(ctx, newNode.EntityID)
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:277]: Instead of querying the entity descriptor you should query the `nodeByEntityKeyFmt` (add state accessor, e.g. `GetEntityNodes`, as currently there is only `HasEntityNodes`). Also expired nodes (`n.IsExpired(epoch)`) should not count against the maximum as those are not eligible to be scheduled.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 4]
```diff
@@ -241,12 +242,71 @@ func (app *registryApplication) registerNode( // nolint: gocyclo
...
+					if grr != nil {
+						ctx.Logger().Error("RegisterNode: failed to query owning entity",
+							"err", grr,
+							"entity", newNode.EntityID,
+						)
+						return grr
+					}
+					// TODO: Do we need to handle entity-signed nodes too?
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:285]: Should be automatically handled if you do it like the above.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 5]
```diff
@@ -573,16 +638,38 @@ func (app *registryApplication) registerRuntime( // nolint: gocyclo
...
+	// Make sure that the entity or runtime has enough stake.
+	// Runtimes using the consensus layer governance model do not require stake.
+	if !params.DebugBypassStake && rt.GovernanceModel != registry.GovernanceConsensus {
 		claim := registry.StakeClaimForRuntime(rt.ID)
 		thresholds := registry.StakeThresholdsForRuntime(rt)
-		acctAddr := staking.NewAddress(rt.EntityID)
+
+		// Figure out which account to use based on the governance model.
+		var acctAddr staking.Address
+		if rt.GovernanceModel == registry.GovernanceEntity {
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:657]: Convert to a switch statement.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 6]
```diff
@@ -573,16 +638,38 @@ func (app *registryApplication) registerRuntime( // nolint: gocyclo
...
+			acctAddr = staking.NewAddress(rt.EntityID)
+		} else if rt.GovernanceModel == registry.GovernanceRuntime {
+			// TODO: I guess we need to wait for #3511 to do this properly :)
+			// We can then probably just use ctx.CallerAddress() in both cases.
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:661]: :+1:

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 7]
```diff
@@ -573,16 +638,38 @@ func (app *registryApplication) registerRuntime( // nolint: gocyclo
...
+			var a signature.PublicKey
+			copy(a[:], rt.ID[:])
+			acctAddr = staking.NewAddress(a)
+		} else {
+			// Basic validation should have caught this, but just in case...
+			ctx.Logger().Error("RegisterRuntime: unknown governance model.")
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:667]:
```suggestion
			ctx.Logger().Error("RegisterRuntime: unknown governance model")
```

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 8]
```diff
@@ -573,16 +638,38 @@ func (app *registryApplication) registerRuntime( // nolint: gocyclo
...
+			return registry.ErrForbidden
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:668]: Should probably be `ErrInvalidArgument`.

File: go/consensus/tendermint/apps/scheduler/scheduler.go [hunk 9]
```diff
@@ -407,7 +409,7 @@ func (app *schedulerApplication) electCommittee(
 		return nil
 	}
 
-	nrNodes, wantedNodes := len(nodeList), workerSize+backupSize
+	nrNodes, wantedNodes := len(nodeList), minSize
```
Comment (kostko) [inline go/consensus/tendermint/apps/scheduler/scheduler.go:412]: That's not what's meant by `MinPoolSize`. There should be an additional check above that fails in case `nrNodes < minSize` (e.g. there should be at least `MinPoolSize` eligible nodes).

File: go/consensus/tendermint/apps/scheduler/scheduler.go [hunk 10]
```diff
@@ -362,7 +362,7 @@ func (app *schedulerApplication) electCommittee(
 		rngCtx       []byte
 		isSuitableFn func(*api.Context, *node.Node, *registry.Runtime) bool
 
-		workerSize, backupSize int
+		workerSize, backupSize, minSize int
```
Comment (kostko) [inline go/consensus/tendermint/apps/scheduler/scheduler.go:365]:
```suggestion
		workerSize, backupSize, minPoolSize int
```

File: go/genesis/genesis_test.go [hunk 11]
```diff
@@ -178,13 +170,13 @@ func TestGenesisSanityCheck(t *testing.T) {
 		},
 		AdmissionPolicy: registry.RuntimeAdmissionPolicy{
 			EntityWhitelist: &registry.EntityWhitelistRuntimeAdmissionPolicy{
-				Entities: map[signature.PublicKey]bool{
-					validPK: true,
+				Entities: map[signature.PublicKey]registry.EntityWhitelistConfig{
+					validPK: {MaxNodes: make(map[node.RolesMask]uint16)},
```
Comment (kostko) [inline go/genesis/genesis_test.go:174]: I think a nil map should behave exactly the same so you could do:
```suggestion
					validPK: {},
```

File: go/oasis-test-runner/scenario/e2e/registry_cli.go [hunk 12]
```diff
@@ -631,11 +632,12 @@ func (sc *registryCLIImpl) testRuntime(ctx context.Context, childEnv *env.Env, c
...
 			MaxApplyWriteLogEntries: 10,
 			MaxApplyOps:             11,
+			MinPoolSize:             9,
 		},
 		AdmissionPolicy: registry.RuntimeAdmissionPolicy{
 			EntityWhitelist: &registry.EntityWhitelistRuntimeAdmissionPolicy{
-				Entities: map[signature.PublicKey]bool{
-					testEntity.ID: true,
+				Entities: map[signature.PublicKey]registry.EntityWhitelistConfig{
+					testEntity.ID: {MaxNodes: make(map[node.RolesMask]uint16)},
```
Comment (kostko) [inline go/oasis-test-runner/scenario/e2e/registry_cli.go:640]: Same as above.

File: go/registry/api/runtime.go [hunk 13]
```diff
@@ -179,10 +187,13 @@ type StorageParameters struct {
...
 	// CheckpointChunkSize is the chunk size parameter for checkpoint creation.
 	CheckpointChunkSize uint64 `json:"checkpoint_chunk_size"`
+
+	// MinPoolSize is the minimum required candidate storage node pool size.
+	MinPoolSize uint64 `json:"min_pool_size"`
 }
 
 // ValidateBasic performs basic storage parameter validity checks.
-func (s *StorageParameters) ValidateBasic() error {
+func (s *StorageParameters) ValidateBasic(ver uint16) error {
```
Comment (kostko) [inline go/registry/api/runtime.go:196]:
```suggestion
func (s *StorageParameters) ValidateBasic(version uint16) error {
```

File: go/registry/api/runtime.go [hunk 14]
```diff
@@ -316,6 +341,56 @@ type Runtime struct { // nolint: maligned
...
+		return gmRuntime
+	case GovernanceConsensus:
+		return gmConsensus
+	default:
+		return "[unsupported runtime governance model]"
+	}
+}
+
+// FromString deserializes a string into a RuntimeGovernanceModel.
+func (gm *RuntimeGovernanceModel) FromString(str string) error {
```
Comment (kostko) [inline go/registry/api/runtime.go:381]: Instead of this implement `MarshalText` and `UnmarshalText` so things like JSON serialization will work as expected.

File: go/registry/api/runtime.go [hunk 15]
```diff
@@ -380,6 +455,11 @@ func (r *Runtime) ValidateBasic(strictVersion bool) error {
 	if err := r.Staking.ValidateBasic(r.Kind); err != nil {
 		return fmt.Errorf("bad staking parameters: %w", err)
 	}
+
+	if v > 1 && (r.GovernanceModel < 1 || r.GovernanceModel > GovernanceConsensus) {
```
Comment (kostko) [inline go/registry/api/runtime.go:459]: Maybe define a constant like `GovernanceMax`.

**Review by kostko — COMMENTED** [review]
Don't forget to update the documentation and add tests for the new features (whitelist changes, minimum pool size, submitting register runtime txes for consensus/runtime governance types).

**Comment by kostko** [comment]
>  add update_runtime runtime message support & e2e tests (maybe in a separate PR?).

Yes, this should be a separate PR.

File: docs/consensus/registry.md [hunk 16]
```diff
@@ -51,12 +51,21 @@ offline and having entities as separate resources enables that.
...
+The chosen governance model indicates how the runtime descriptor can be updated
+in the future.
+There are currently three supported governance models:
+    * **entity governance** -- the runtime owner is the only one who can update
+      the runtime descriptor via `registry.RegisterRuntime` method calls.
+    * **runtime-defined governance** -- the runtime itself is the only one who
+      can update the runtime descriptor by emitting a runtime message.
+    * **consensus layer governance** -- only the consensus layer itself can
+      update the runtime descriptor either through a network upgrade or some
+      other future means.
```
Comment (kostko) [inline docs/consensus/registry.md:68]:
```suggestion
The chosen governance model indicates how the runtime descriptor can be updated
in the future.

There are currently three supported governance models:

* **Entity governance** where the runtime owner is the only one who can update
  the runtime descriptor via `registry.RegisterRuntime` method calls.

* **Runtime-defined governance** where the runtime itself is the only one who
  can update the runtime descriptor by emitting a runtime message.

* **Consensus layer governance** where only the consensus layer itself can
  update the runtime descriptor through network governance.
```

File: go/consensus/tendermint/apps/registry/state/state.go [hunk 17]
```diff
@@ -443,6 +375,46 @@ func (s *ImmutableState) NodeStatus(ctx context.Context, id signature.PublicKey)
...
+		if !signedNodeByEntityKeyFmt.Decode(it.Key(), &hEntityID, &hNodeID) || !hEntityID.Equal(&hID) {
+			break
+		}
+
+		rawSignedNode, err := s.is.Get(ctx, signedNodeKeyFmt.Encode(&hNodeID))
+		if err != nil {
+			return nil, abciAPI.UnavailableStateError(err)
+		}
+		if rawSignedNode == nil {
+			return nil, registry.ErrNoSuchNode
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/state/state.go:400]: Doesn't this indicate state corruption as if the node is removed the `signedNodeByEntityKey` index should also be updated? In this case wrap it in `abciAPI.UnavailableStateError(...)`.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 18]
```diff
@@ -241,13 +241,74 @@ func (app *registryApplication) registerNode( // nolint: gocyclo
...
 			ctx.Logger().Error("RegisterNode: node's entity not in a runtime's whitelist",
 				"entity", newNode.EntityID,
 				"runtime", rt.ID,
 			)
 			return registry.ErrForbidden
 		}
+		if wcfg.MaxNodes == nil {
+			continue
+		}
+		if len(wcfg.MaxNodes) == 0 {
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:258]: This should also cover the `nil` case so you don't need to have two checks.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 19]
```diff
@@ -241,13 +241,74 @@ func (app *registryApplication) registerNode( // nolint: gocyclo
...
+			continue
+		}
+
+		// Map is present and non-empty, check per-role restrictions
+		// on the maximum number of nodes per entity.
+
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:264]: The below code needs some test coverage.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 20]
```diff
@@ -542,13 +602,19 @@ func (app *registryApplication) registerRuntime( // nolint: gocyclo
...
 	// NOTE: If this is invoked during InitChain then there is no actual transaction
 	//       and thus no transaction signer so we must skip this check.
-	if !ctx.IsInitChain() && !sigRt.Signature.PublicKey.Equal(ctx.TxSigner()) {
+	if !ctx.IsInitChain() && !rt.EntityID.Equal(ctx.TxSigner()) {
 		return registry.ErrIncorrectTxSigner
 	}
 
+	// Consensus-layer governance runtimes can only be registered during genesis.
+	if !ctx.IsInitChain() && rt.GovernanceModel == registry.GovernanceConsensus {
+		return registry.ErrForbidden
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:615]: Add test.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 21]
```diff
@@ -573,16 +639,33 @@ func (app *registryApplication) registerRuntime( // nolint: gocyclo
 		if err != nil {
 			return err
 		}
+		// If the governance model is runtime-defined governance, the owner
+		// cannot perform any updates after the initial registration.
+		// Updates to such runtimes should be done via runtime messages.
+		if existingRt.GovernanceModel == registry.GovernanceRuntime {
+			return registry.ErrForbidden
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:646]: Add test.

File: go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go [hunk 22]
```diff
@@ -137,14 +222,64 @@ func doFixGenesis(cmd *cobra.Command, args []string) {
...
+		KeyManager:      rt.KeyManager,
+		Staking:         rt.Staking,
+		TxnScheduler:    rt.TxnScheduler,
+		GovernanceModel: registry.GovernanceEntity,
+		Executor: registry.ExecutorParameters{
+			GroupSize:         rt.Executor.GroupSize,
+			GroupBackupSize:   rt.Executor.GroupBackupSize,
+			AllowedStragglers: rt.Executor.AllowedStragglers,
+			RoundTimeout:      rt.Executor.RoundTimeout,
+			MaxMessages:       32, // TODO: What should this be set to?
```
Comment (kostko) [inline go/oasis-node/cmd/debug/fixgenesis/fixgenesis.go:242]: Good question, so setting this also requires configuring the global roothash consensus parameter (`MaxRuntimeMessages`). Either fixgenesis should set that as well (to the same value?) or we leave this at zero. Let's set this to 32 (both per-runtime and the global roothash consensus parameter) and then we can discuss what to set this to before the actual upgrade.

File: go/oasis-node/cmd/registry/runtime/runtime.go [hunk 23]
```diff
@@ -405,17 +405,22 @@ func runtimeFromFlags() (*registry.Runtime, signature.Signer, error) { // nolint
...
-				return nil, nil, fmt.Errorf("entity whitelist runtime admission policy parse entity ID: %w", err)
+				return nil, fmt.Errorf("entity whitelist runtime admission policy parse entity ID: %w", err)
 			}
-			entities[e] = true
+			entities[e] = registry.EntityWhitelistConfig{
+				MaxNodes: make(map[node.RolesMask]uint16),
+			}
+			// TODO: Handle the clusterfuck of parsing the nested map from
+			// command-line arguments sometime later.  As it is now, it's
+			// configured as unlimited nodes of any role for the given entity.
```
Comment (kostko) [inline go/oasis-node/cmd/registry/runtime/runtime.go:423]: Don't forget to file an issue.

File: go/registry/api/api.go [hunk 24]
```diff
@@ -995,66 +984,53 @@ func exactlyOneTrue(conds ...bool) bool {
...
+		return fmt.Errorf("%w: test runtime not allowed", ErrInvalidArgument)
 	}
 
 	if err := rt.Genesis.SanityCheck(isGenesis); err != nil {
-		return nil, err
+		return err
+	}
+
+	// Make sure the specified runtime governance model is allowed.
+	if params.EnableRuntimeGovernanceModels == nil {
```
Comment (kostko) [inline go/registry/api/api.go:1019]:
```suggestion
	if len(params.EnableRuntimeGovernanceModels) == 0 {
```

File: go/registry/api/api.go [hunk 25]
```diff
@@ -995,66 +984,53 @@ func exactlyOneTrue(conds ...bool) bool {
...
+		// No runtime governance models are allowed.
+		return fmt.Errorf("%w: no runtime governance models are enabled", ErrForbidden)
+	}
+	if allowed, exists := params.EnableRuntimeGovernanceModels[rt.GovernanceModel]; !exists || !allowed {
```
Comment (kostko) [inline go/registry/api/api.go:1023]: The default value in case a key doesn't exist is `false` so just do:
```suggestion
	if !params.EnableRuntimeGovernanceModels[rt.GovernanceModel] {
```

File: go/registry/api/api.go [hunk 26]
```diff
@@ -1165,6 +1172,23 @@ func VerifyRuntimeUpdate(logger *logging.Logger, currentRt, newRt *Runtime) erro
...
+	// Check if governance model update is valid.
+	if currentRt.GovernanceModel != newRt.GovernanceModel {
+		// Transitioning from entity to runtime governance is allowed, but
+		// all other transitions are not.
+		if !(currentRt.GovernanceModel == GovernanceEntity && newRt.GovernanceModel == GovernanceRuntime) {
+			logger.Error("RegisterRuntime: invalid governance model transition",
+				"current_gm", currentRt.GovernanceModel.String(),
+				"new_gm", newRt.GovernanceModel.String(),
+			)
+			return ErrRuntimeUpdateNotAllowed
```
Comment (kostko) [inline go/registry/api/api.go:1184]: Add test.

File: go/registry/api/runtime.go [hunk 27]
```diff
@@ -110,6 +113,11 @@ func (e *ExecutorParameters) ValidateBasic() error {
 	if e.RoundTimeout < 5 {
 		return fmt.Errorf("round timeout too small")
 	}
+
+	if version > 1 && e.MinPoolSize < e.GroupSize+e.GroupBackupSize {
+		return fmt.Errorf("minimum pool size too small")
```
Comment (kostko) [inline go/registry/api/runtime.go:118]: Add test.

File: go/registry/api/runtime.go [hunk 28]
```diff
@@ -267,7 +292,7 @@ func (s *RuntimeStakingParameters) ValidateBasic(runtimeKind RuntimeKind) error
 const (
 	// LatestRuntimeDescriptorVersion is the latest entity descriptor version that should be used
 	// for all new descriptors. Using earlier versions may be rejected.
-	LatestRuntimeDescriptorVersion = 1
+	LatestRuntimeDescriptorVersion = 2
 
 	// Minimum and maximum descriptor versions that are allowed.
 	minRuntimeDescriptorVersion = 1
```
Comment (kostko) [inline go/registry/api/runtime.go:298]: When would we support version 1 runtime descriptors? Isn't the idea that fixgenesis will convert them and so we don't need to support version 1 here (so no need for version checks above and below)?

File: go/registry/api/runtime.go [hunk 29]
```diff
@@ -316,6 +341,61 @@ type Runtime struct { // nolint: maligned
...
+		return gmRuntime
+	case GovernanceConsensus:
+		return gmConsensus
+	default:
+		return "[unsupported runtime governance model]"
+	}
+}
+
+func (gm RuntimeGovernanceModel) MarshalText() ([]byte, error) {
+	return []byte(gm.String()), nil
```
Comment (kostko) [inline go/registry/api/runtime.go:383]: This shouldn't just use `.String()` as that can in theory return the `[unsupported runtime governance model]`. It should instead return an error in this case.

File: go/staking/api/slashing.go [hunk 30]
```diff
@@ -41,6 +41,10 @@ func (s SlashReason) MarshalText() ([]byte, error) {
 // UnmarshalText decodes a text slice into a SlashReason.
 func (s *SlashReason) UnmarshalText(text []byte) error {
 	switch string(text) {
+	// XXX: The "0" case is only for backward compatibility, so that the old
+	// genesis file loads -- remove this once mainnet is upgraded!
```
Comment (kostko) [inline go/staking/api/slashing.go:45]: Wouldn't it be better if we handled this in fixgenesis?

File: go/consensus/tendermint/apps/registry/state/state.go (hunk 17 above)
Comment (abukosek) [reply go/consensus/tendermint/apps/registry/state/state.go:400]: Good catch, I'll add that.

File: go/consensus/tendermint/apps/registry/transactions.go (hunk 19 above)
Comment (abukosek) [reply go/consensus/tendermint/apps/registry/transactions.go:264]: Yeah, I'm adding that now :)

File: go/registry/api/runtime.go (hunk 28 above)
Comment (abukosek) [reply go/registry/api/runtime.go:298]: Oops, this is a leftover from before I wrote the fixgenesis code, I'll change the minimum version to 2.

File: go/staking/api/slashing.go (hunk 30 above)
Comment (abukosek) [reply go/staking/api/slashing.go:45]: No, because there are too many structures to duplicate and there's more chance of something going wrong with the conversion (and since this is in staking code it's better to be more careful).  A two-line fix like this is cleaner, IMO :)
I'll file an issue later, so we don't forget to remove this once the upgrade is complete.

**Comment by abukosek** [comment]
Thanks for the comments!  This is now ready for re-review.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 31]
```diff
@@ -542,13 +599,19 @@ func (app *registryApplication) registerRuntime( // nolint: gocyclo
 		return err
 	}
 
-	// Make sure the signer of the transaction matches the signer of the runtime.
+	// Make sure the signer of the transaction matches the signer of the
+	// entity that is controlling the runtime.
 	// NOTE: If this is invoked during InitChain then there is no actual transaction
 	//       and thus no transaction signer so we must skip this check.
-	if !ctx.IsInitChain() && !sigRt.Signature.PublicKey.Equal(ctx.TxSigner()) {
+	if !ctx.IsInitChain() && !rt.EntityID.Equal(ctx.TxSigner()) {
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:606]: This actually needs to take into account the existing governance model. So this check is true for entity governance, but for runtime governance the check would look something like checking `ctx.CallerAddress()` for equality with `staking.NewRuntimeAddress(rt.ID)`.

So maybe put a switch here that decides based on the governance model. The switch could also incorporate the conditional below which returns `ErrForbidden` in case of consensus governance.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 32]
```diff
@@ -241,13 +241,71 @@ func (app *registryApplication) registerNode( // nolint: gocyclo
...
+		// only have a single role).
+		for _, role := range node.Roles() {
+			if !newNode.HasRoles(role) {
+				// Skip unset roles.
+				continue
+			}
+
+			maxNodes, exists := wcfg.MaxNodes[role]
+			if !exists {
+				// No such role found in whitelist.
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:272]: Log something in these cases (below) that return `ErrForbidden`.

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 33]
```diff
@@ -573,16 +650,33 @@ func (app *registryApplication) registerRuntime( // nolint: gocyclo
 		if err != nil {
 			return err
 		}
+		// If the governance model is runtime-defined governance, the owner
+		// cannot perform any updates after the initial registration.
+		// Updates to such runtimes should be done via runtime messages.
+		if existingRt.GovernanceModel == registry.GovernanceRuntime {
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:656]: If the runtime update message will also call `registerRuntime` then this check will prevent any updates from happening. In case the governance model is the same as in `rt` then this has already been checked in the switch above (checking against `ctx.CallerAddress()`). So the only remaining issue is if the governance model in `rt` would be different than in `existingRt` -- but that should be checked in `VerifyRuntimeUpdate`?

Comment (abukosek) [reply go/consensus/tendermint/apps/registry/transactions.go]: You are right, I've fixed this now.  I've also moved the signer checks below this code, so we can do them based on the existing runtime governance model instead of the new one (otherwise it wouldn't be possible to transition from entity to runtime governance, since the transaction for doing that will still be signed with the entity's signer and not the runtime's) :)

File: go/consensus/tendermint/apps/registry/transactions.go (hunk 31 above)
Comment (abukosek) [reply go/consensus/tendermint/apps/registry/transactions.go]: Thanks, fixed! :)

File: go/registry/api/runtime.go [hunk 34]
```diff
@@ -179,10 +187,13 @@ type StorageParameters struct {
...
 	// CheckpointChunkSize is the chunk size parameter for checkpoint creation.
 	CheckpointChunkSize uint64 `json:"checkpoint_chunk_size"`
+
+	// MinPoolSize is the minimum required candidate storage node pool size.
+	MinPoolSize uint64 `json:"min_pool_size"`
 }
 
 // ValidateBasic performs basic storage parameter validity checks.
-func (s *StorageParameters) ValidateBasic() error {
+func (s *StorageParameters) ValidateBasic(version uint16) error {
```
Comment (kostko) [inline go/registry/api/runtime.go:196]: The version argument should no longer be needed as we only allow version 2 descriptors.

File: go/registry/api/runtime.go [hunk 35]
```diff
@@ -214,6 +225,11 @@ func (s *StorageParameters) ValidateBasic() error {
 			return fmt.Errorf("storage CheckpointChunkSize parameter too small")
 		}
 	}
+
+	if version > 1 && s.MinPoolSize < s.GroupSize {
```
Comment (kostko) [inline go/registry/api/runtime.go:229]:
```suggestion
	if s.MinPoolSize < s.GroupSize {
```

File: go/registry/api/runtime.go [hunk 36]
```diff
@@ -380,6 +464,11 @@ func (r *Runtime) ValidateBasic(strictVersion bool) error {
 	if err := r.Staking.ValidateBasic(r.Kind); err != nil {
 		return fmt.Errorf("bad staking parameters: %w", err)
 	}
+
+	if v > 1 && (r.GovernanceModel < 1 || r.GovernanceModel > GovernanceMax) {
```
Comment (kostko) [inline go/registry/api/runtime.go:468]:
```suggestion
	if r.GovernanceModel < 1 || r.GovernanceModel > GovernanceMax {
```

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 37]
```diff
@@ -575,14 +638,67 @@ func (app *registryApplication) registerRuntime( // nolint: gocyclo
...
+		var rtToCheck *registry.Runtime
+		if existingRt != nil {
+			rtToCheck = existingRt
+		} else {
+			rtToCheck = rt
+		}
+
+		switch rtToCheck.GovernanceModel {
+		case registry.GovernanceEntity:
+			if !rt.EntityID.Equal(ctx.TxSigner()) {
```
Comment (kostko) [inline go/consensus/tendermint/apps/registry/transactions.go:660]: I know the EntityID cannot change but this makes it more clear IMO:
```suggestion
			if !rtToCheck.EntityID.Equal(ctx.TxSigner()) {
```

File: go/consensus/tendermint/apps/registry/transactions.go (hunk 33 above)
Comment (kostko) [reply go/consensus/tendermint/apps/registry/transactions.go]: Looks good now, thanks!

File: docs/consensus/registry.md [hunk 38]
```diff
@@ -51,12 +51,24 @@ offline and having entities as separate resources enables that.
...
-identifier, kind, admission policy, committee scheduling, storage etc. For a
-full description of the runtime descriptor see [the `Runtime` structure].
+identifier, kind, admission policy, committee scheduling, storage, governance
+model, etc. For a full description of the runtime descriptor see
+[the `Runtime` structure].
 
-Currently only the owning entity is allowed to make any modifications to the
-runtime. There are plans to enable runtimes to update their own descriptors in
-the future to enable runtimes to be self-governing.
+<!-- markdownlint-disable no-space-in-emphasis -->
```
[the remark this answers is not in the document]
Comment (abukosek) [reply docs/consensus/registry.md:58]: Yeah, it seems broken by default -- it complains about the spaces in emphasised phrases.
I've re-enabled it below now :)

File: go/consensus/tendermint/apps/registry/transactions.go [hunk 39]
```diff
@@ -241,13 +241,84 @@ func (app *registryApplication) registerNode( // nolint: gocyclo
...
+			if grr != nil {
+				ctx.Logger().Error("RegisterNode: failed to query entity nodes",
+					"err", grr,
+					"entity", newNode.EntityID,
+				)
+				return grr
+			}
+			var curNodes uint16
+			for _, n := range nodes {
+				if n.IsExpired(uint64(epoch)) || n.GetRuntime(rt.ID) == nil {
```
[the remark this answers is not in the document]
Comment (abukosek) [reply go/consensus/tendermint/apps/registry/transactions.go]: Good catch!  I've fixed it now.

File: go/consensus/tendermint/apps/registry/genesis.go [hunk 40]
```diff
@@ -46,47 +46,43 @@ func (app *registryApplication) InitChain(ctx *abciAPI.Context, request types.Re
...
-					"runtime", v,
+					"runtime_id", rt.ID,
 				)
 				return fmt.Errorf("registry: genesis runtime registration failure: %w", err)
 			}
 		}
 	}
-	for i, v := range st.SuspendedRuntimes {
-		if v == nil {
+	for i, rt := range st.SuspendedRuntimes {
```
[the remark this answers is not in the document]
Comment (abukosek) [reply go/consensus/tendermint/apps/registry/genesis.go:72]: Filed as https://github.com/oasisprotocol/oasis-core/issues/3619 :)

File: docs/consensus/registry.md (hunk 38 above)
[the remark this answers is not in the document]
Comment (ptrus) [reply docs/consensus/registry.md:58]: thx!

**Review by kostko — APPROVED** [review]

File: go/oasis-node/cmd/registry/runtime/runtime.go (hunk 23 above)
Comment (abukosek) [reply go/oasis-node/cmd/registry/runtime/runtime.go:423]: Filed as https://github.com/oasisprotocol/oasis-core/issues/3626 :)

File: go/staking/api/slashing.go (hunk 30 above)
Comment (abukosek) [reply go/staking/api/slashing.go:45]: Filed as https://github.com/oasisprotocol/oasis-core/issues/3627 :)

**Comment by abukosek** [comment]
Thanks!
